### PR TITLE
Implement accessibility for tizen platform

### PR DESCRIPTION
**Pull request by wanchao-xu**
- Transplant linux part of chromium accessibility.
  - Screen reader reads widget information by atk/atspi2 on tizne platform, and chromium accessibility support atk/atspi2
  - Screen reader only read highlighted widget information on tizen mobile/wearable, modify the platform node of chromium accessibility to support it.
- Implement the accessibility for tizen platform
  - To use the flutter tizen accessibility, the screen reader should apply the patch https://review.tizen.org/gerrit/#/c/platform/core/accessibility/screen-reader/+/270107/.

### CODE REVIEW

**Comment by swift-kim** [comment]
Problem 1) The engine is being included as part of the embedder binary.

You can clearly see that the size of the embedder is larger than that of the engine, which is unexpected.

```
swift@swift-760XDA:~/Git/engine/src$ ls -lh out/linux_release_arm/lib*.so
-rwxrwxr-x 1 swift swift 7.8M  3월  4 13:31 out/linux_release_arm/libflutter_engine.so
-rwxrwxr-x 1 swift swift 8.2M  3월  4 13:31 out/linux_release_arm/libflutter_tizen_common.so
-rwxrwxr-x 1 swift swift 8.2M  3월  4 13:31 out/linux_release_arm/libflutter_tizen_mobile.so
-rwxrwxr-x 1 swift swift 8.2M  3월  4 13:31 out/linux_release_arm/libflutter_tizen_tv.so
-rwxrwxr-x 1 swift swift 8.2M  3월  4 13:31 out/linux_release_arm/libflutter_tizen_wearable.so
```

Here's a temporary solution:

```patch
diff --git a/shell/platform/common/BUILD.gn b/shell/platform/common/BUILD.gn
index 74c200d90a..1bbf2c8eb8 100644
--- a/shell/platform/common/BUILD.gn
+++ b/shell/platform/common/BUILD.gn
@@ -89,9 +89,13 @@ source_set("common_cpp_accessibility") {

   public_deps = [
     "//flutter/fml:fml",
-    "//flutter/shell/platform/embedder:embedder_as_internal_library",
+    "//flutter/shell/platform/embedder:embedder_headers",
     "//flutter/third_party/accessibility",
   ]
+
+  if (!build_tizen_shell) {
+    public_deps += [ "//flutter/shell/platform/embedder:embedder_as_internal_library" ]
+  }
 }

 source_set("common_cpp") {
```

Problem 2) Non-public API usage detected:

```
Symbols not allowed (out/linux_release_arm/libflutter_tizen_wearable.so):
 U atk_action_get_type
 U atk_bridge_adaptor_init
 U atk_component_get_type
 U atk_document_get_type
 U atk_hyperlink_get_type
 U atk_hyperlink_impl_get_type
 U atk_hypertext_get_type
 U atk_image_get_type
 U atk_object_get_name
 U atk_object_get_type
 U atk_object_initialize
 U atk_object_notify_state_change
 U atk_relation_set_add_relation_by_type
 U atk_relation_set_new
 U atk_relation_type_get_type
 U atk_selection_get_type
 U atk_state_set_add_state
 U atk_table_get_type
 U atk_text_attribute_get_name
 U atk_text_get_character_count
 U atk_text_get_type
 U atk_util_get_type
 U atk_value_get_type
 U atk_window_get_type
 U kFlutterSemanticsCustomActionIdBatchEnd
 U kFlutterSemanticsNodeIdBatchEnd
```

This means that apps built with the embedder binary will be automatically rejected by the watch app store. You may need to either exclude the code (related to the accessibility feature) from the wearable embedder build (by modifying `tizen/BUILD.gn` and using `#ifndef WEARABLE_PROFILE`) or send a request to add the symbols to the store's API allowlist.

**Comment by wanchao-xu** [comment]
@swift-kim , about "Problem 2) Non-public API usage detected", wearable support accessibility(screen reader), and if I exclude the code, screen reader will not read the widget information of flutter app, I think it's better to add the symbols to the store's API allowlist.

**Comment by swift-kim** [comment]
Please format the GN file (`shell/platform/common/BUILD.gn`) correctly so that the CI build can run.

> I think it's better to add the symbols to the store's API allowlist.

By the way, do you know why the symbols used by the embedder itself (`kFlutterSemanticsCustomActionIdBatchEnd`, `kFlutterSemanticsNodeIdBatchEnd`) are externally exported? We should avoid adding these symbols to the allowlist as much as possible.

**Comment by wanchao-xu** [comment]
about kFlutterSemanticsCustomActionIdBatchEnd and kFlutterSemanticsNodeIdBatchEnd, flutter engine will get all semantics node and custom actions, the symbols are used to notice the end.

```
bool FlutterTizenEngine::RunEngine(const char* entrypoint) {
  ...
  args.update_semantics_node_callback = OnUpdateSemanticsNode;
  args.update_semantics_custom_action_callback = OnUpdateSemanticsCustomActions;
```

**Comment by swift-kim** [comment]
> about kFlutterSemanticsCustomActionIdBatchEnd and kFlutterSemanticsNodeIdBatchEnd

Anyway you need to find a way to prevent these symbols from being exported.

> And I have no list of atk_* symbols, is there any way to show such list?

You may take a look at the `src/build/linux/debian_sid_amd64-sysroot/debian/libatk1.0-0/DEBIAN/symbols` file in your buildroot or https://gnome.pages.gitlab.gnome.org/atk/api-index-full.html. According to https://github.com/flutter-tizen/tizen_tools/runs/5418572124?check_suite_focus=true, you are using the libatk version 2.30.0.

**Comment by swift-kim** [comment]
@wanchao-xu Where did you copy the Chromium Linux accessibility code from? I need to be able to reproduce what you've done in order to upgrade the engine and accessibility code in the future. I need very detailed information on each file (e.g. why `ax_platform_node_base.cc` has been modified) but not just a link to the repo.

**Comment by wanchao-xu** [comment]
I copied the chromium linux accessibility from https://chromium.googlesource.com/chromium/src/+/4579d5538f06c5ef615a15bc67ebb9ac0523a973. Please refer to `src/flutter/third_party/accessibility/README.md`/.

The original chromium accessibility has many dependencies, I should modify some code to reduce the dependencies after transplanted. Google also modify the chromium accessibility to support mac and windows, not just a link to the repo.

**Comment by wanchao-xu** [comment]
To use the flutter tizen accessibility, the screen reader should apply the patch https://review.tizen.org/gerrit/#/c/platform/core/accessibility/screen-reader/+/270107/.

File: third_party/accessibility/ax/platform/ax_platform_node_base.cc [hunk 1]
```diff
@@ -1764,6 +1764,91 @@ bool AXPlatformNodeBase::IsPlatformCheckable() const {
...
+      return *boundary_offset;
+  }
+
+  // std::vector<int32_t> unused_line_start_offsets;
+  // return static_cast<int>(
+  //     FindAccessibleTextBoundary(GetHypertext(), unused_line_start_offsets,
+  //                                boundary, offset, direction, affinity));
+  BASE_LOG() << "FindTextBoundary of SentenceStart is not yet supported.";
+  BASE_UNREACHABLE();
+  return offset;
```
Comment (swift-kim) [inline third_party/accessibility/ax/platform/ax_platform_node_base.cc:1849]: Is this an unnecessary feature?

File: third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    return nullptr;
+
+  int action;
+  if (!obj->GetIntAttribute(ax::mojom::IntAttribute::kDefaultActionVerb,
+                            &action))
+    return nullptr;
+  std::string action_verb =
+      ui::ToLocalizedString(static_cast<ax::mojom::DefaultActionVerb>(action));
+  ATK_AURALINUX_RETURN_STRING(action_verb);
+}*/
```
Comment (swift-kim) [inline third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc:221]: Why has this been commented out?

File: third_party/accessibility/ax/ax_node_position.cc [hunk 3]
```diff
@@ -14,7 +14,7 @@
 namespace ui {
 
 AXEmbeddedObjectBehavior g_ax_embedded_object_behavior =
-#if defined(OS_WIN)
+#if defined(OS_WIN) || defined(OS_LINUX)
     AXEmbeddedObjectBehavior::kExposeCharacter;
 #else
     AXEmbeddedObjectBehavior::kSuppressCharacter;
```
Comment (swift-kim) [inline third_party/accessibility/ax/ax_node_position.cc:20]: Is this change necessary for Tizen?

File: third_party/accessibility/ax/platform/ax_platform_node_base.cc (hunk 1 above)
Comment (wanchao-xu) [reply third_party/accessibility/ax/platform/ax_platform_node_base.cc:1849]: I'm not sure it's necessary or not. The `FindAccessibleTextBoundary` will refer to many dependencies which are exclude in flutter, and I checked the other code, the `ax::mojom::TextBoundary::kSentenceStart' is not supported in line 1617 of ax_position.h. 

```
      case ax::mojom::TextBoundary::kSentenceStart:
        BASE_LOG() << "Sentence boundaries are not yet supported.";
        BASE_UNREACHABLE();
        return CreateNullPosition();
```
I think we can uncomment this code in future when we got the scene to use it.

Comment (swift-kim) [reply third_party/accessibility/ax/platform/ax_platform_node_base.cc:1849]: Okay, thanks :)

File: third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc (hunk 2 above)
Comment (wanchao-xu) [reply third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc:221]: I commented this code out in order to fix the build error, `ui::ToLocalizedString` should use `l10n` in Chromium's `base/l10n`.
Screen reader use `get_name` and `get_description` to read information usually. It's no matter to commented this code out.

```
static void atk_action_interface_init(AtkActionIface* iface) {
  iface->do_action = ax_platform_atk_hyperlink_do_action;
  iface->get_n_actions = ax_platform_atk_hyperlink_get_n_actions;
  iface->get_description = ax_platform_atk_hyperlink_get_description;
  iface->get_keybinding = ax_platform_atk_hyperlink_get_keybinding;
  iface->get_name = ax_platform_atk_hyperlink_get_name;
  // iface->get_localized_name = ax_platform_atk_hyperlink_get_localized_name;
}
```

File: third_party/accessibility/base/string_utils.cc [hunk 4]
```diff
@@ -4,6 +4,7 @@
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply third_party/accessibility/base/string_utils.cc:4]: I will change the permission of this file.

File: third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc (hunk 2 above)
Comment (swift-kim) [reply third_party/accessibility/ax/platform/ax_platform_atk_hyperlink.cc:221]: Okay, thanks. I can also see the build error when the code is uncommented.

File: shell/platform/tizen/BUILD.gn [hunk 5]
```diff
@@ -117,7 +118,9 @@ template("embedder") {
       "external_texture_surface_gl.cc",
       "flutter_project_bundle.cc",
       "flutter_tizen.cc",
+      "flutter_tizen_accessibility_bridge_delegate.cc",
       "flutter_tizen_engine.cc",
+      "flutter_tizen_platform_node_delegate.cc",
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/BUILD.gn]: I changed the classes to `AccessibilityBridgeDelegateTizen` and `FlutterPlatformNodeDelegateTizen`.

File: shell/platform/tizen/accessibility_settings.cc [hunk 6]
```diff
@@ -18,7 +14,23 @@
 namespace flutter {
 
 AccessibilitySettings::AccessibilitySettings(FlutterTizenEngine* engine)
-    : engine_(engine) {
+    : engine_(engine), screen_reader_enabled_(false) {
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_settings.cc]: I modified `screen_reader_enabled_` initialization in [88fcd99](https://github.com/flutter-tizen/engine/pull/254/commits/88fcd99a7db6e527347630d62511a627edfd2c7a).
This variable is needed, when I enable tts on mobile simulator, the `OnScreenReaderStateChanged` may be called multiple times sometimes, so I add this variable to prevent `SetSemanticsEnabled` repeatedly.

File: shell/platform/tizen/accessibility_settings.cc [hunk 7]
```diff
@@ -18,7 +14,23 @@
...
+      screen_reader_enabled_ = enabled;
+      engine_->SetSemanticsEnabled(enabled);
+    }
+  } else {
+    FT_LOG(Error) << "Failed to get value of accessibility tts.";
+  }
+
+  // add listener for accessibility tts
+  system_settings_set_changed_cb(SYSTEM_SETTINGS_KEY_ACCESSIBILITY_TTS,
+                                 OnScreenReaderStateChanged, this);
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_settings.cc]: I changed `OnHighContrastStateChanged` as `OnScreenReaderStateChanged ` in [88fcd99](https://github.com/flutter-tizen/engine/pull/254/commits/88fcd99a7db6e527347630d62511a627edfd2c7a)

File: shell/platform/tizen/accessibility_settings.cc [hunk 8]
```diff
@@ -38,12 +50,34 @@ AccessibilitySettings::AccessibilitySettings(FlutterTizenEngine* engine)
...
+  system_settings_unset_changed_cb(SYSTEM_SETTINGS_KEY_ACCESSIBILITY_TTS);
 #ifdef TV_PROFILE
   system_settings_unset_changed_cb(system_settings_key_e(
       SYSTEM_SETTINGS_KEY_MENU_SYSTEM_ACCESSIBILITY_HIGHCONTRAST));
 #endif
 }
 
+bool AccessibilitySettings::IsAccessibilityEnabled() {
+  return screen_reader_enabled_;
+}
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_settings.cc]: I removed this api in [88fcd99](https://github.com/flutter-tizen/engine/pull/254/commits/88fcd99a7db6e527347630d62511a627edfd2c7a).

File: shell/platform/tizen/accessibility_settings.cc [hunk 9]
```diff
@@ -18,7 +14,23 @@
...
 AccessibilitySettings::AccessibilitySettings(FlutterTizenEngine* engine)
-    : engine_(engine) {
+    : engine_(engine), screen_reader_enabled_(false) {
+  bool enabled = false;
+  int result = system_settings_get_value_bool(
+      SYSTEM_SETTINGS_KEY_ACCESSIBILITY_TTS, &enabled);
+  if (result == SYSTEM_SETTINGS_ERROR_NONE) {
+    if (enabled) {
+      screen_reader_enabled_ = enabled;
+      engine_->SetSemanticsEnabled(enabled);
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_settings.cc]: `engine_` is nerver null, I modified it in [88fcd99](https://github.com/flutter-tizen/engine/pull/254/commits/88fcd99a7db6e527347630d62511a627edfd2c7a).

File: shell/platform/tizen/accessibility_settings.h [hunk 10]
```diff
@@ -1,10 +1,12 @@
-// Copyright 2022 Samsung Electronics Co., Ltd. All rights reserved.
+// Copyright 2020 Samsung Electronics Co., Ltd. All rights reserved.
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_settings.h]: Fixed it in [b83f9b0](https://github.com/flutter-tizen/engine/pull/254/commits/b83f9b0b621ab2895b96f0dd52c40c47a38fb994)

File: shell/platform/tizen/channels/accessibility_channel.cc [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+              auto data = std::get<flutter::EncodableMap>(eData);
+              auto& eMsg = data[flutter::EncodableValue("message")];
+              if (std::holds_alternative<std::string>(eMsg)) {
+                std::string msg = std::get<std::string>(eMsg);
+                elm_atspi_bridge_utils_say(msg.c_str(), true, nullptr, nullptr);
+              }
+            }
+            reply(flutter::EncodableValue());
+          }
+        }
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc]: I fixed it in [02c09cd](https://github.com/flutter-tizen/engine/pull/254/commits/02c09cde7773ff982ec639dc3e4c66ef3e038c17)

File: shell/platform/tizen/channels/accessibility_channel.cc [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  channel_->SetMessageHandler([&](const auto& message, auto reply) {
+    if (enabled_) {
+      if (std::holds_alternative<flutter::EncodableMap>(message)) {
+        auto eMessage = std::get<flutter::EncodableMap>(message);
+        auto& eType = eMessage[flutter::EncodableValue("type")];
+        auto& eData = eMessage[flutter::EncodableValue("data")];
+        if (std::holds_alternative<std::string>(eType)) {
+          std::string type = std::get<std::string>(eType);
+          FT_LOG(Info) << "Recevie " << type
+                       << " event from accessibility channel";
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc]: I fixed it in [02c09cd](https://github.com/flutter-tizen/engine/pull/254/commits/02c09cde7773ff982ec639dc3e4c66ef3e038c17)

File: shell/platform/tizen/channels/accessibility_channel.cc [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+AccessibilityChannel::AccessibilityChannel(BinaryMessenger* messenger,
+                                           bool enabled)
+    : enabled_(enabled),
+      channel_(std::make_unique<BasicMessageChannel<EncodableValue>>(
+          messenger,
+          kChannelName,
+          &StandardMessageCodec::GetInstance())) {
+  channel_->SetMessageHandler([&](const auto& message, auto reply) {
+    if (enabled_) {
+      if (std::holds_alternative<flutter::EncodableMap>(message)) {
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc]: I removed `enabled_` in [ef38a5f](https://github.com/flutter-tizen/engine/pull/254/commits/ef38a5fc72ee76bf1f33692c00da0b9135dea19c)

**Review by swift-kim — COMMENTED** [review]
Just minor style issues.

The code generally looks good to me but I haven't tested on a real device.

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.h [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+  // |FlutterPlatformNodeDelegate|
+  gfx::NativeViewAccessible GetParent() override;
+
+  // |FlutterPlatformNodeDelegate|
+  gfx::Rect GetBoundsRect(
+      const ui::AXCoordinateSystem coordinate_system,
+      const ui::AXClippingBehavior clipping_behavior,
+      ui::AXOffscreenResult* offscreen_result) const override;
+
+  bool HitTestPoint(const gfx::Point& point) const;
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.h:32]: `HitTestPoint` is also used in `FlutterPlatformWindowDelegateTizen`

```
gfx::NativeViewAccessible FlutterPlatformWindowDelegateTizen::HitTestSync(
    int screen_physical_pixel_x,
    int screen_physical_pixel_y) const {
  gfx::Point point(screen_physical_pixel_x, screen_physical_pixel_y);
  if (!root_.expired()) {
    auto node = std::static_pointer_cast<FlutterPlatformNodeDelegateTizen>(
        root_.lock());
    if (node->HitTestPoint(point)) {
      return node->HitTestSync(screen_physical_pixel_x,
                               screen_physical_pixel_y);
    }
  }
```
This function should be public.

[the remark this answers is not in the document]
Comment (swift-kim) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.h:32]: Oh, I didn't notice that they were different classes. Thanks.

File: shell/platform/tizen/channels/accessibility_channel.cc [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+          kChannelName,
+          &StandardMessageCodec::GetInstance())) {
+  channel_->SetMessageHandler([&](const auto& message, auto reply) {
+    if (std::holds_alternative<EncodableMap>(message)) {
+      auto map = std::get<EncodableMap>(message);
+      EncodableValueHolder<std::string> type(&map, "type");
+      EncodableValueHolder<EncodableMap> data(&map, "data");
+
+      FT_LOG(Info) << "Received " << *type << " message.";
+      if (type && *type == "announce" && data) {
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc:33]: `if (type && *type == "announce" && data) {`, the `type` does null check with `if(type)` before `*type`.

File: shell/platform/tizen/accessibility_bridge_delegate_tizen.cc [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  auto bridge = engine_->GetAccessibilityBridge().lock();
+  if (!bridge) {
+    FT_LOG(Error) << "Accessibility bridge is deallocated";
+    return;
+  }
+
+  if (!targeted_event.node) {
+    FT_LOG(Error) << "AXNode of targeted event is null";
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_bridge_delegate_tizen.cc]: I removed it in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/accessibility_bridge_delegate_tizen.cc [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+      targeted_event.event_params.event);
+}
+
+void AccessibilityBridgeDelegateTizen::DispatchAccessibilityAction(
+    AccessibilityNodeId target,
+    FlutterSemanticsAction action,
+    fml::MallocMapping data) {
+  if (engine_) {
+    engine_->DispatchAccessibilityAction(target, action, std::move(data));
+  }
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_bridge_delegate_tizen.cc]: I removed the null check in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/accessibility_bridge_delegate_tizen.h [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+#define EMBEDDER_ACCESSIBILITY_BRIDGE_DELEGATE_TIZEN_H_
+
+#include "flutter/shell/platform/common/accessibility_bridge.h"
+
+namespace flutter {
+
+class FlutterTizenEngine;
+
+// The tizen implementation of AccessibilityBridge::AccessibilityBridgeDelegate.
+// This delegate is used to create AccessibilityBridge in the tizen platform.
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/accessibility_bridge_delegate_tizen.h]: I capitalized `t` of all string `tizen` in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355)

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.cc [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+void FlutterPlatformNodeDelegateTizen::NotifyAccessibilityEvent(
+    ui::AXEventGenerator::Event event_type) {
+  if (!platform_node_) {
+    FT_LOG(Error) << "Platform node isn't created";
+    return;
+  }
+
+  auto paltform_node =
+      static_cast<ui::AXPlatformNodeAuraLinux*>(platform_node_);
+  FT_LOG(Debug) << "handel event " << ui::ToString(event_type) << " on node "
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.cc]: I fixed the string mistake in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.cc [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+  return bounds;
+}
+
+bool FlutterPlatformNodeDelegateTizen::HitTestPoint(
+    const gfx::Point& point) const {
+  gfx::Rect bounds =
+      GetBoundsRect(ui::AXCoordinateSystem::kScreenPhysicalPixels,
+                    ui::AXClippingBehavior::kUnclipped, nullptr);
+  return point.x() >= bounds.x() && point.x() <= bounds.right() &&
+         point.y() >= bounds.y() && point.y() <= bounds.bottom();
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.cc]: `Containes` is replaced in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.cc [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+FlutterPlatformAppDelegateTizen::~FlutterPlatformAppDelegateTizen() {
+  platform_node_->Destroy();
+  platform_node_ = nullptr;
+}
+
+// static
+FlutterPlatformAppDelegateTizen&
+FlutterPlatformAppDelegateTizen::GetInstance() {
+  static FlutterPlatformAppDelegateTizen TizenAppInstance;
+  return TizenAppInstance;
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.cc]: I changed the variable name with `instance` in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/channels/accessibility_channel.cc [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+namespace flutter {
+
+namespace {
+
+constexpr char kChannelName[] = "flutter/accessibility";
+
+}  // namespace
+
+AccessibilityChannel::AccessibilityChannel(BinaryMessenger* messenger,
+                                           bool enabled)
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc]: The `enabled` is removed in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 23]
```diff
@@ -560,4 +565,77 @@ FlutterRendererConfig FlutterTizenEngine::GetRendererConfig() {
   return config;
 }
 
+std::weak_ptr<AccessibilityBridge>
+FlutterTizenEngine::GetAccessibilityBridge() {
+  return accessibility_bridge_;
+}
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_tizen_engine.cc]: I modified this api like 'key_event_channel` in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 24]
```diff
@@ -560,4 +565,77 @@ FlutterRendererConfig FlutterTizenEngine::GetRendererConfig() {
...
+
+void FlutterTizenEngine::OnUpdateSemanticsNode(const FlutterSemanticsNode* node,
+                                               void* user_data) {
+  if (node->id == kFlutterSemanticsNodeIdBatchEnd) {
+    return;
+  }
+
+  FT_LOG(Debug) << "Update semantics node [id=" << node->id
+                << " label=" << node->label << " hint=" << node->hint
+                << " value=" << node->value << "]";
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_tizen_engine.cc]: I added the commas between the values, and also applied to line 632. Please refer to [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

File: shell/platform/tizen/flutter_tizen_engine.cc [hunk 25]
```diff
@@ -560,4 +565,77 @@ FlutterRendererConfig FlutterTizenEngine::GetRendererConfig() {
...
+  auto engine = reinterpret_cast<FlutterTizenEngine*>(user_data);
+  if (engine->accessibility_bridge_) {
+    engine->accessibility_bridge_->AddFlutterSemanticsNodeUpdate(node);
+  } else {
+    FT_LOG(Error) << "Accessibility bridge must be initialized. ";
```
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_tizen_engine.cc]: The space is removed in [ce07326](https://github.com/flutter-tizen/engine/pull/254/commits/ce07326011f2e4e2cf2b956e0fe4978a84282355).

**Comment by swift-kim** [comment]
> I think it's better to add the symbols to the store's API allowlist.

@wanchao-xu I discussed with Mr. Kwon today and decided not to enable accessibility support for wearable devices (mainly due to MX division's recent security issue).

You can apply this patch to disable the feature from the wearable embedder: https://github.com/swift-kim/engine/commit/76c43c8e8edd24a3162bafde7e2ad449fe9e5056. (Please cherry-pick the commit but do not manually modify the code to avoid human errors.)

Please make a suggestion if you have a better idea.

File: shell/platform/tizen/channels/accessibility_channel.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (swift-kim) [reply shell/platform/tizen/channels/accessibility_channel.cc:33]: I mean, it should be like this:

```suggestion
      if (type && *type == "announce" && data) {
        FT_LOG(Info) << "Received " << *type << " message.";
```

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.cc (hunk 20 above)
[the remark this answers is not in the document]
Comment (swift-kim) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.cc]: Can this be applied to `flutter_platform_node_delegate_tizen.cc`'s line 248-249 as well?

File: shell/platform/tizen/channels/accessibility_channel.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/channels/accessibility_channel.cc]: The log is printed after null check of `type`. Please refer to [4ac1994](https://github.com/flutter-tizen/engine/pull/254/commits/4ac1994be13212f1f30155c2ab280b66c8901b8f).

File: shell/platform/tizen/flutter_platform_node_delegate_tizen.cc (hunk 20 above)
[the remark this answers is not in the document]
Comment (wanchao-xu) [reply shell/platform/tizen/flutter_platform_node_delegate_tizen.cc]: I fixed it in [4ac1994](https://github.com/flutter-tizen/engine/pull/254/commits/4ac1994be13212f1f30155c2ab280b66c8901b8f).

**Review by swift-kim — APPROVED** [review]
I tested the feature with Voice Guide enabled on a debug TV and it worked great. Thank you for your hard work!

**Comment by wanchao-xu** [comment]
@swift-kim , Can I merge this request into branch now?
